### PR TITLE
HashMath::Mapper Algorithm

### PR DESCRIPTION
**Pull request by mattruggio**
This PR adds a new algorithm that can perform constant-time lookup using pre-materialized datasets to add/update a hashes key-value pairs in order to fill in missing data.  Take a look at the README and class documentation for examples.

### CODE REVIEW

File: lib/hash_math/unpivot.rb [hunk 1]
```diff
@@ -13,7 +13,6 @@ module HashMath
   # This class has the ability to extrapolate one hash (row) into multiple hashes (rows) while
   # unpivoting specific keys into key-value pairs.
   class Unpivot
-    acts_as_hashable
```
Comment (mattruggio) [inline lib/hash_math/unpivot.rb:16]: random cleanup

File: lib/hash_math/unpivot/pivot.rb [hunk 2]
```diff
@@ -34,7 +34,7 @@ def initialize(keys:, coalesce_key:, coalesce_key_value:)
       # the first will serve as the prototype for each returned hash
       # the second will be one to use for value extraction.
       # Returns an array of hashes.
-      def expand(base_hash, value_hash)
```
Comment (mattruggio) [inline lib/hash_math/unpivot/pivot.rb:37]: random cleanup

File: lib/hash_math/unpivot/pivot_set.rb [hunk 3]
```diff
@@ -32,7 +32,7 @@ def add(pivot)
 
       # An aggregation of Pivot#expand.  This method will iterate over all pivots
       # and expand them all out.
-      def expand(hash)
```
Comment (mattruggio) [inline lib/hash_math/unpivot/pivot_set.rb:35]: random cleanup

File: lib/hash_math/mapper/mapping.rb [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    # ----------------------------------------------------------------------------------------------
+    #
+    # mapped_patient now equals: { id: 1, code: 'active', patient_status_id: 1 }
+    class Mapping
+      extend Forwardable
+      acts_as_hashable
+
+      attr_reader :value, :set, :with, :lookup
+
+      def_delegators :lookup, :name, :by, :add, :add_each
```
Comment (ddewar) [inline lib/hash_math/mapper/mapping.rb:36]: I believe :add and :add_each are duplicates of methods on this class.

Comment (mattruggio) [reply lib/hash_math/mapper/mapping.rb]: Good find, removing delegation.

File: lib/hash_math/mapper.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    # to initialize.
+    def initialize(mappings = [])
+      mappings          = Mapping.array(mappings)
+      @mappings_by_name = pivot_by_name(mappings)
+
+      freeze
+    end
+
+    # Add an enumerable list of lookup records to this instance's lookup dataset.
+    def add_each(name, objects)
```
[the remark this answers is not in the document]
Comment (mattruggio) [reply lib/hash_math/mapper.rb:30]: I agree, it does not make sense when name is blank, I will add guard clauses.

**Pull request merged**